ecl/hqlcpp/hqlttcpp.cpp

```diff
@@ -6224,18 +6224,81 @@ IHqlExpression * WorkflowTransformer::transformInternalFunction(IHqlExpression *
 
     HqlExprArray funcdefArgs;
     funcdefArgs.append(*LINK(newBody));
-    unwindChildren(funcdefArgs, newFuncDef, 1);
-    OwnedHqlExpr namedFuncDef = newFuncDef->clone(funcdefArgs);
-    inheritDependencies(namedFuncDef);
 
     if (ecl->getOperator() == no_embedbody)
+    {
+        IHqlExpression * outofline = newFuncDef->queryChild(0);
+        IHqlExpression * formals = newFuncDef->queryChild(1);
+        IHqlExpression * defaults = newFuncDef->queryChild(2);
+        assertex(outofline->getOperator() == no_outofline);
+        IHqlExpression * bodyCode = outofline->queryChild(0);
+        HqlExprArray attrArgs;
+        ForEachChild(idx, bodyCode)
+        {
+            IHqlExpression *child = bodyCode->queryChild(idx);
+            if (child->isAttribute() && child->queryName() != languageAtom && child->queryName() != importAtom)
+            {
+                StringBuffer attrParam;
+                if (attrArgs.ordinality())
+                    attrParam.append(",");
+                attrParam.append(child->queryName());
+
+                IHqlExpression * value = child->queryChild(0);
+                if (value)
+                    attrParam.append("=");
+                attrArgs.append(*createConstant(attrParam));
+                if (value)
+                    attrArgs.append(*ensureExprType(value, unknownStringType));
+            }
+        }
+        OwnedHqlExpr folded;
+        if (attrArgs.length())
+        {
+            OwnedHqlExpr concat = createUnbalanced(no_concat, unknownStringType, attrArgs);
+            OwnedHqlExpr cast = ensureExprType(concat, unknownVarStringType);
+
+            // It's not legal to use parameters in the options, since it becomes ambiguous whether they should be bound to embed variables or not.
+            // Check that they didn't and give a sensible error message
+            OwnedHqlExpr boundCast = replaceInlineParameters(newFuncDef, cast);
+            if (cast != boundCast)
+                throwError(HQLERR_EmbedParamNotSupportedInOptions);
+
+            folded.setown(foldHqlExpression(cast));
+        }
+        else
+            folded.setown(createConstant(""));
+        HqlExprArray newFormals;
+        HqlExprArray attrs;
+        unwindChildren(newFormals, formals, 0);
+        newFormals.append(*createParameter(createIdAtom("__options"), newFormals.length(), LINK(unknownVarStringType), attrs));
+
+        HqlExprArray newDefaults;
+        if (defaults)
+            unwindChildren(newDefaults, defaults, 0);
+        while (newDefaults.length() < formals->numChildren())
+            newDefaults.append(*createOmittedValue());
+        newDefaults.append(*folded.getClear());
+        formals = formals->clone(newFormals);
+        defaults = createValueSafe(no_sortlist, makeSortListType(NULL), newDefaults);
+        funcdefArgs.append(*LINK(formals));
```

Comment:
That may start getting complicated.  I suspect this transformation should actually be done in HqlTreeNormalizer instead.